ipam/claims.go

```diff
@@ -0,0 +1,90 @@
+package ipam
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/zettio/weave/router"
+)
+
+type pendingClaim struct {
+	resultChan chan<- error
+	Ident      string
+	IP         net.IP
+}
+
+type claimList []pendingClaim
+
+func (aa *claimList) removeAt(pos int) {
+	(*aa) = append((*aa)[:pos], (*aa)[pos+1:]...)
+}
+
+func (aa *claimList) find(addr net.IP) int {
+	for i, a := range *aa {
+		if a.IP.Equal(addr) {
+			return i
+		}
+	}
+	return -1
+}
+
+// Claim an address that we think we should own
+func (alloc *Allocator) handleClaim(ident string, addr net.IP, resultChan chan<- error) {
+	if !alloc.ring.Contains(addr) {
+		// Address not within our universe; assume user knows what they are doing
+		alloc.infof("Ignored address %s claimed by %s - not in our universe", addr, ident)
+		resultChan <- nil
+		return
+	}
+	// See if it's already claimed
+	if pos := alloc.claims.find(addr); pos >= 0 && alloc.claims[pos].Ident != ident {
+		resultChan <- fmt.Errorf("IP address %s already claimed by %s", addr, alloc.claims[pos].Ident)
+		return
+	}
+	alloc.infof("Address %s claimed by %s", addr, ident)
+	if owner, err := alloc.checkClaim(ident, addr); err != nil {
+		resultChan <- err
+	} else if owner != alloc.ourName {
+		alloc.infof("Address %s owned by %s", addr, owner)
+		alloc.claims = append(alloc.claims, pendingClaim{resultChan, ident, addr})
+	} else {
+		resultChan <- nil
+	}
+}
```

Comment:
inline this function at the call site. Yes, it's quite big, but the special treatment of resultChan is too obscure otherwise.